### PR TITLE
CPM-556: Add Product Groups user intents

### PR DESCRIPTION
**Pull request by Balamung**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/AddGroupsApplier.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function apply(UserIntent $groupsUserIntent, ProductInterface $product, int $userId): void
+    {
+        Assert::isInstanceOf($groupsUserIntent, AddGroups::class);
+
+        $existingGroups = $product->getGroupCodes();
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/AddGroupsApplier.php:32]: I prefer "formerGroups" naming.
Existing is more "the group exists in the PIM"

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/RemoveGroupsApplier.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function apply(UserIntent $groupsUserIntent, ProductInterface $product, int $userId): void
+    {
+        Assert::isInstanceOf($groupsUserIntent, RemoveGroups::class);
+
+        $existingGroups = $product->getGroupCodes();
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/RemoveGroupsApplier.php:32]: $formerGroups

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/Groups/RemoveGroupsApplierSpec.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        $this->shouldHaveType(RemoveGroupsApplier::class);
+        $this->shouldImplement(UserIntentApplier::class);
+    }
+
+    function it_applies_a_remove_groups_user_intent_on_a_product(
+        ObjectUpdaterInterface $productUpdater,
+        Product $product
+    ) {
+        $userIntent = new RemoveGroups(['promotion']);
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/Groups/RemoveGroupsApplierSpec.php:34]: Is there something specific to do when one of the remove/add command has an empty array ? 
like new RemoveGroups([])
Should we throw something ? (ask nini, I think no)

**Review by pierallard — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/Groups/RemoveGroupsApplierSpec.php (hunk 3 above)
Comment (Balamung) [reply src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/Groups/RemoveGroupsApplierSpec.php:34]: I think if that's the case, we should put it directly in an Assert in the intent's constructor ? @nmarniesse what do you think ? :)

Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/Groups/RemoveGroupsApplierSpec.php:34]: yes in the intent constructor

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/AddGroupsApplier.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        Assert::isInstanceOf($groupsUserIntent, AddGroups::class);
+
+        $formerGroups = $product->getGroupCodes();
+        $newGroups = \array_values(\array_unique(\array_merge($formerGroups, $groupsUserIntent->groupCodes())));
+
+        if (\count(\array_diff($newGroups, $formerGroups)) > 0) {
+            $this->productUpdater->update($product, [
+                'groups' => $newGroups,
+            ]);
+        }
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/AddGroupsApplier.php:39]: easier IMO:
```php
$formerGroupCodes = $product->getGroupCodes();
if ([] === \array_diff($groupsUserIntent->groupCodes(), $formerGroupCodes)) {
    return;
}
$this->productUpdater->update($product, [
    'groups' => \array_values(\array_unique(\array_merge($formerGroupCodes, $groupsUserIntent->groupCodes()))),
]);

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/RemoveGroupsApplier.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        Assert::isInstanceOf($groupsUserIntent, RemoveGroups::class);
+
+        $formerGroups = $product->getGroupCodes();
+        $newGroups = \array_values(\array_unique(\array_diff($formerGroups, $groupsUserIntent->groupCodes())));
+
+        if (\count(\array_diff($formerGroups, $newGroups)) > 0) {
+            $this->productUpdater->update($product, [
+                'groups' => $newGroups,
+            ]);
+        }
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/RemoveGroupsApplier.php:39]:
```suggestion
        $formerGroupCodes = \array_values($product->getGroupCodes());
        $updatedGroupCodes = \array_diff($formerGroupCodes, $groupsUserIntent->groupCodes());
        if ($formerGroupCodes === $updatedGroupCodes) {
            return;
        }

        $this->productUpdater->update($product, [
            'groups' => \array_values($updatedGroupCodes),
        ]);
```

File: src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/Groups/AddGroups.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Akeneo\Pim\Enrichment\Product\API\Command\UserIntent\Groups;
+
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class AddGroups implements GroupsUserIntent
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/Groups/AddGroups.php:13]: AddToGroups (the business user intent here is really adding a product to a group, not adding a group to a product)

File: src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/Groups/RemoveGroups.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Akeneo\Pim\Enrichment\Product\API\Command\UserIntent\Groups;
+
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class RemoveGroups implements GroupsUserIntent
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/Groups/RemoveGroups.php:13]: RemoveFromGroups

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/SetGroupsApplier.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function apply(UserIntent $groupsUserIntent, ProductInterface $product, int $userId): void
+    {
+        Assert::isInstanceOf($groupsUserIntent, SetGroups::class);
+
+        $this->productUpdater->update($product, [
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/SetGroupsApplier.php:32]: add early return if former and new groups are the same

**Review by mmetayer — COMMENTED** [review]
Is validation of the groups' existence planned in another card?

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/SetGroupsApplier.php (hunk 8 above)
Comment (Balamung) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/Groups/SetGroupsApplier.php:36]: sort both and then compare

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/Groups/SetGroupsApplierSpec.php [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        ObjectUpdaterInterface $productUpdater
+    ) {
+        $userIntent = new SetGroups(['promotion', 'toto']);
+        $product = new Product();
+        $product->setIdentifier('foo');
+
+        $productUpdater->update($product, ['groups' => ['promotion', 'toto']])->shouldBeCalledOnce();
+
+        $this->apply($userIntent, $product, 10);
+    }
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Application/Applier/Groups/SetGroupsApplierSpec.php:38]: Add test if same groups with different order

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/Groups/AddToGroupsSpec.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class AddToGroupsSpec extends ObjectBehavior
+{
+    function let()
+    {
+        $this->beConstructedWith(['promotions', 'toto']);
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/Groups/AddToGroupsSpec.php:19]: Check that asserts are triggered correctly

**Review by nmarniesse — APPROVED** [review]

**Review by mmetayer — APPROVED** [review]
GTM once naming is fixed

**Pull request merged**